File: LeetCode/Python/#206 Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

#Iteration
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        pre = None

        while head:
            tmp = head.next
            head.next = pre
            pre = head
            head = tmp

        return pre

# Recursion
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        return self.reverse(head)

    def reverse(self,node,pre=None):
        if not node:
            return pre
        n = node.next
        node.next = pre
        return self.reverse(n,node)

File: LeetCode/JavaScript/#206. Reverse Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// Iteration
var reverseList = function(head) {
     var pre = null;
     var tmp = null;

     while(head){
     	tmp = head.next;
     	head.next = pre;
     	pre = head;
     	head = tmp;
     }
     return pre;
};

// recursion
var reverseList = function(head){
	return head?reversed(head,null):head;
}

function reversed(node,pre){
	let newList = node;

	if(node.next){
		newList = reversed(node.next,node);
	}
	node.next = pre;

	return newList;
}